src/devices/machine/gayle.cpp

```diff
@@ -69,117 +86,317 @@ void gayle_device::device_start()
+	map(0xb000, 0xb001).rw(FUNC(gayle_device::control_r), FUNC(gayle_device::control_w));
 }
 
 
 //**************************************************************************
-//  IMPLEMENTATION
+//  INTERNAL
 //**************************************************************************
 
-uint16_t gayle_device::gayle_r(offs_t offset, uint16_t mem_mask)
+void gayle_device::dump_register()
+{
+	LOGMASKED(LOG_REG, "status = %02x, change = %02x, int = %02x, control = %02x\n",
+		m_gayle_reg[REG_STATUS], m_gayle_reg[REG_CHANGE], m_gayle_reg[REG_INT], m_gayle_reg[REG_CONTROL]);
+}
+
+void gayle_device::line_change(int line, int state, int level)
 {
-	uint16_t data = 0xffff;
-	offset <<= 1;
+	LOGMASKED(line == LINE_IDE ? LOG_IDE : LOG_CC, "line_change: %d %d %d\n", line, state, level);
 
-	// swap
-	mem_mask = swapendian_int16(mem_mask);
+	m_line_state &= ~(1 << line);
+	m_line_state |= (state << line);
 
-	if (BIT(offset, 15))
+	// ignore pcmcia line changes if the interface is disabled
+	if (line != LINE_IDE && BIT(m_gayle_reg[REG_STATUS], 0))
+		return;
+
+	// did we change state?
+	if (BIT(m_gayle_reg[REG_STATUS], line) != state)
 	{
-		switch (offset & 0x7fff)
+		// indicate a change to the line
+		m_gayle_reg[REG_CHANGE] |= 1 << line;
+
+		// special handling for line 6 (credit card detect)
+		if (line == LINE_CC_DET && BIT(m_gayle_reg[REG_CHANGE], 1))
 		{
-		case 0x0000: data = m_gayle_reg[0]; break;
-		case 0x1000: data = m_gayle_reg[1]; break;
-		case 0x2000: data = m_gayle_reg[2]; break;
-		case 0x3000: data = m_gayle_reg[3]; break;
+			LOGMASKED(LOG_GENERAL, "resetting due to credit card detection change\n");
+
+			m_rst_w(0);
+			m_rst_w(1);
+			reset();
+
+			return;
 		}
-	}
-	else
-	{
-		if (!BIT(offset, 14))
+
+		if (state)
+			m_gayle_reg[REG_STATUS] |= 1 << line;
+		else
+			m_gayle_reg[REG_STATUS] &= ~(1 << line);
+
+		// generate interrupt (if enabled)
+		if (BIT(m_gayle_reg[REG_INT], line))
 		{
-			if (BIT(offset, 13))
-				data = m_cs0_read((offset >> 2) & 0x07, mem_mask);
+			LOGMASKED(line == LINE_IDE ? LOG_IDE : LOG_CC, "line %d int assert\n", line);
+
+			if (level == 2)
+				m_int2_w(1);
 			else
-				data = m_cs1_read((offset >> 2) & 0x07, mem_mask);
+				m_int6_w(1);
 		}
 	}
+}
+
 
-	if (VERBOSE)
-		logerror("gayle_r(%06x): %04x & %04x\n", offset, data, mem_mask);
+//**************************************************************************
+//  REGISTER
+//**************************************************************************
 
-	// swap data
-	data = swapendian_int16(data);
+uint16_t gayle_device::status_r()
+{
+	// 7-------  ide interrupt status
+	// -6------  credit card detect
+	// --5-----  battery voltage 1/status change
+	// ---4----  battery voltage 2/digital audio
+	// ----3---  credit card write enable
+	// -----2--  credit card busy/interrupt request
+	// ------1-  enable pcmcia digital audio
+	// -------0  disable pcmcia
+
+	LOGMASKED(LOG_REG, "status_r: %02x\n", m_gayle_reg[REG_STATUS]);
```

Comment:
This will get very noisy if  a debugger memory window is open.  You should guard the log message with `if (!machine().side_effects_disabled())` so you only get log messages for actual reads.